Standard API qps and burst in kubeconfig file. (#1629)
kubernetes/enhancements: keps/sig-api-machinery/1630-kubeconfig-qps-burst/README.md

-->

- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [ ] KEP approvers have approved the KEP status as `implementable`
- [ ] Design details are appropriately documented
- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] Graduation criteria is in place
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Kubernetes REST clients use a simple token bucket rate limiter to control the pace of requests from a client.

> We will expose the values already present in the [REST config](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/config.go#L114-L120), via a kubeconfig file inside the client stanza.

Reviewer: "via a kubeconfig file inside the client stanza" sounds like it is referring to a kubeconfig file that is inside some client stanza --- but I do not think this is what's meant here.  I think what's meant is more like "via the user definitions in kubeconfig files".  Alternatively, perhaps what is meant is more like "via a kubeconfig file that is added to a ServiceAccount's Secret".

Reviewer 2: ```suggestion
We will expose the values already present in the [REST config](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/config.go#L114-L120), in the client stanza of a kubeconfig file.
```

Reviewer: you mean "a user stanza"?

We will wire the in-cluster kubeconfig file to honor the presence of a new file to allow customization inside a cluster at the discretion of a different, user provided, controller.

## Motivation

We've always had reasons to override these values, so we created one-off arguments (`--kube-api-burst`, `--kube-api-qps`).
With widespread adoption of operators and other in-cluster workloads, having a standard way to provide these values makes sense.
In addition, with the introduction of API priority and fairness, being able to configure and even disable this client-side rate limiting in compatible environments is important.

### Goals

> 1. Allow easy and consistent configuration for the existing client-side rate limiting.

Reviewer 2: Is this our goal, or is our goal to get rid of it completely?

Is this necessary only transitionally, or forever?

### Non-Goals

1. Allow customization of client-side rate limiting strategies.

## Proposal

> Add to the `AuthInfo` section of a kubeconfig.

Reviewer: Why not in the `preferences`?  That would merge just fine and not involve the complicated (and poorly documented) logic involved in merging users --- which does not even meet the need here.

Author: > Why not in the `preferences`? That would merge just fine and not involve the complicated (and poorly documented) logic involved in merging users --- which does not even meet the need here.

That's a dead field.

This section identifies a particular client identity.
This identity spans all the contexts with different namespaces, so it is a closer logical match.
You don't generally want different QPS characteristics based on namespace.

> ```go

Reviewer: Please include a URL or k/k-relative pathname for the file where this mod would apply.

type AuthInfo struct {
	// ...
	// QPS indicates the maximum QPS to the master from this client.

> 	// If it's zero, the created RESTClient will use DefaultQPS: 5

Reviewer 3: should we clarify a negative QPS will turn off the ratelimiting in the doc?

Reviewer 4: "Set to -1 to turn off rate limiting. A value of 0 means to use the default limit from the client binary."

Reviewer 3: -1 can be even confusing given that the QPS is a float32. what's supposed to happen it's other negative values than -1?

Reviewer 2: Agree with others, clarify how to turn it off, clarify what negative numbers do (if not that).

I don't really want to make this super more complicated than necessary, but making this into a union/one-of block might make sense.

	QPS float32

	// Maximum burst for throttle.
	// If it's zero, the created RESTClient will use DefaultBurst: 10.
	Burst int32
}
```

> ### in-cluster kubeconfig overrides

Reviewer: The vision here is not entirely clear.  I suspect the idea is that the Secret of a ServiceAccount would be extended to have a data member named `overrides.kubeconfig` containing a kubeconfig file that specifies the QPS and Burst and nothing else, and that anyone who wants to exercise this control has to create a controller that modifies all the ServiceAccount Secret objects.  Have I got this right?  If so, that's pretty painful to use and requires a controller with extreme privilege.

Author: > Have I got this right? If so, that's pretty painful to use and requires a controller with extreme privilege.

Yeah, but anything we do is going to require mutation of that same secret.  It isn't to be taken lightly in general.  This is about making it possible, I don't think easy is needed.

Reviewer 2: This QPS limit has to be a property of the service account, so this does seem necessary.

Maybe it's not the only model though, I'll make another comment.

The in-cluster config is the automatic configuration that exists within pods.
The client code itself can be written to consume an overrides.kubeconfig file, which will be merged with the standard

> kubeconfig merging rules that we use the env var path today.

Reviewer: The merging rules we have today do not allow adding/modifying fields within a user.  The first specification of that user is taken as a whole and no part is overridden.

Author: > The merging rules we have today do not allow adding/modifying fields within a user. The first specification of that user is taken as a whole and no part is overridden.

No, we merge all of it:  https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/loader.go#L226-L252 

You're probably used to selection, but merging is also a thing.

Reviewer: Yes, I looked at that code.  You see I raised issues with the poor comment on that function (https://github.com/kubernetes/kubernetes/issues/91092) and the poor user-facing documentation about this (https://github.com/kubernetes/website/issues/20967).

The code in question does not itself answer the question, it just makes a two-handed pile of invocations of `mergo.MergeWithOverwrite`.  The comments on that suggest that merging will be done, but the implementation does not do that.  The implementation does not simply look at each layer of structure independently.  In the case of a map, the behavior depends on the type of each value.  In the case of a map entry whose value is a struct or pointer, the implementation gets to https://github.com/imdario/mergo/blob/v0.3.5/merge.go#L143 --- which replaces the whole value without recursing into it.

I tested, and indeed found that when merging two kubeconfig files that define different fields for a certain user, the result is only the fields of the first kubeconfig file to define that user.

The controller code will *not* be updated.
If a cluster-admin wants to control this file, he can provide his own controller to write content into this file in 
serviceaccount token secrets.

### Test Plan

This is heavily unit testable.

### Graduation Criteria

The API itself is well-known since near the beginning of kube.

> We are simply exposing existing values, it will enter as GA.

Reviewer: There is significant novelty here, in the override mechanism.

Author: > There is significant novelty here, in the override mechanism.

I don't think so.  I built this pre-1.0 so the merging code already exists.  This is just about the expression of what merges.

Reviewer: Perhaps the merging behavior is different today.

### Upgrade / Downgrade Strategy

Old clients will ignore the unknown field.
Old `kubectl` will rewrite the stanza and remove the new values.

### Version Skew Strategy

Old clients will ignore the unknown field.
Old `kubectl` will rewrite the stanza and remove the new values.

## Implementation History

> ## Drawbacks

Reviewer 2: Do other, non-go clients parse kubeconfig files?

I assume repeated invocations of kubectl would not respect the limits in aggregate?

Reviewer: I think this is aimed at long-lived clients rather than kubectl.  I would not expect client-side rate limits to be relevant to kubectl.

Reviewer 2: I would not expect that either, but users who see the settings in their kubeconfig files might.

Reviewer: This proposal _allows_ new content in the kubeconfig files that users hold, but does not require it in general nor in the motivating scenario in particular.

> ## Alternatives

Reviewer 2: Another alternative might be to plumb these via env var. That'd be gross, but very direct.

Reviewer: Yeah.

Reviewer 2: The go team tends to do things this way, e.g. for disabling their home grown DNS resolver, and also for disabling HTTP2 I think.

The benefit of this is that it only requires one component to change, no API is baked in, we can remove the env var without breaking people who are setting it.

The downside is that it's not very discoverable (but maybe that's OK since we mostly want to use it to run tests).

Reviewer 2: Anyone setting this should also look at the client's metrics to verify that it's not spending any time blocked on the QPS limiter (I think there's already a metric).

1. Connecting to the kube-apiserver to retrieve a value for default QPS and burst.
   No default value is going to be good across all workloads.

>    It's also mechanicaly weird to create a client to get a client.

Reviewer 2: I agree this is a bit weird, but it wouldn't have to be a separate request-- we could make apiserver set a header with its desired rate limit, and the client just obeys it after the first request. This might be simpler?

Reviewer 2: I am actually liking this (return a header) more the longer I think about it. Since we have to recompile all clients anyway...

Reviewer: Yeah.

Reviewer 2: After more thought I think this is too heavyweight (on API changes -- but so is the main proposal), and it's easy to believe that not all clients would want the same rate limit.

Reviewer 2: (So I guess this would still be in consideration if it were just about turning off the rate limit, not about setting it.)
